<a href="https://colab.research.google.com/github/CJLucido/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Learning Objectives

- Objective 01 - understand the purpose of feature engineering
- Objective 02 - demonstrate how to work with strings in pandas
- Objective 03 - modify or create dataframe columns using the `apply()` function
- Objective 04 - work with dates and times in pandas


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective 01](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity, and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [ ]:
import pandas as pd

# Pandas Display Options:
pd.set_option("display.max_rows", 150)
pd.set_option("display.max_columns", 100)

In [ ]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

### Specific Columns hold specific kinds of data

In [ ]:
print(df.shape)
df.head()

(1460, 81)


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [ ]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [ ]:
# Look at a few rows of the `BedroomAbvGr` column.
# Looks like integers to me!
df['BedroomAbvGr'].head()


0    3
1    3
2    3
3    3
4    4
Name: BedroomAbvGr, dtype: int64

What type of variable is BedroomAbvGr?  Is it categorical or quantitative?  If you answered "categorical", is it ordinal, nominal or an identifier?  If you answered "quantitative", is it continuous or discrete?

It is quantitative and discrete.

Some columns hold float values like the `LotFrontage` column.

In [ ]:
# Look at a few rows of the `LotFrontage` column.
df['LotFrontage'].tail()


1455    62.0
1456    85.0
1457    66.0
1458    68.0
1459    75.0
Name: LotFrontage, dtype: float64

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [ ]:
df['LotFrontage']


0       65.0
1       80.0
2       68.0
3       60.0
4       84.0
        ... 
1455    62.0
1456    85.0
1457    66.0
1458    68.0
1459    75.0
Name: LotFrontage, Length: 1460, dtype: float64

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [ ]:
# NaN is the most common value in this column. What is a NaN
df['LotFrontage'].value_counts(dropna=False).head()

NaN     259
60.0    143
70.0     70
80.0     69
50.0     57
Name: LotFrontage, dtype: int64

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes sense to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember - Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [ ]:
import numpy as np

# What is the datatype of NaN?
type(np.NaN)


float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice`


In [ ]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
# get a single column
small_df = df[['TotalBsmtSF','1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()
small_df.head()


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0          856       856       854     208500
1         1262      1262         0     181500
2          920       920       866     223500
3          756       961       756     140000
4         1145      1145      1053     250000

In [ ]:
# Get more than one column, then pass in a list of column headers



In [ ]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets



In [ ]:
small_df.dtypes

### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [ ]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

# Using bracket syntax to make a new 'TotalSquareFootage' column
small_df['TotalSquareFootage'] = small_df['1stFlrSF'] + small_df['2ndFlrSF'] + small_df['TotalBsmtSF']
small_df.head()


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0          856       856       854     208500                2566
1         1262      1262         0     181500                2524
2          920       920       866     223500                2706
3          756       961       756     140000                2473
4         1145      1145      1053     250000                3343

In [ ]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

small_df['PricePerSqFt'] = small_df['SalePrice'] / small_df['TotalSquareFootage']
small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  
0     81.254871  
1     71.909667  
2     82.594235  
3     56.611403  
4     74.783129

###We can also use if-then statements to create new variables.  

Say we want to categorize houses as having a high price per square foot (greater than or equal to 80 dollars per square foot) or a low price per square foot (less than 80 dollars per square foot).

In [ ]:
def is_it_high(cell_value):
  if cell_value > 80:
    return True
  else:
    return False

small_df['High_ppsqft'] = small_df['PricePerSqFt'].apply(is_it_high)#he did this with a for loop but I already knew about the apply method
small_df.head()
#pd.crosstab(small_df['High_ppsqft'],columns='count')

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  High_ppsqft  
0     81.254871         True  
1     71.909667        False  
2     82.594235         True  
3     56.611403        False  
4     74.783129        False

Now we have added several new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



### Let's include "and" and "or" conditions in the if-then statements.

Let's identify the 2-story duplex houses.

## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective 02](#work-with-strings) - Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [ ]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-10-01 23:35:00--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 54.213.174.123, 44.235.65.250, 44.224.91.33
Connecting to resources.lendingclub.com (resources.lendingclub.com)|54.213.174.123|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip.1’

LoanStats_2018Q4.cs     [             <=>    ]  22.30M  2.68MB/s    in 9.6s    

2020-10-01 23:35:10 (2.32 MB/s) - ‘LoanStats_2018Q4.csv.zip.1’ saved [23387713]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [ ]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [ ]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [ ]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","2287.33","2287.33","4364.28","4364.28","3312.67","1051.61","0.0","0.0","0.0","Sep-2020","190.21","Oct-2020","Sep-2020","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [ ]:
# Read in the CSV
#import pandas as pd

df = pd.read_csv('LoanStats_2018Q4.csv', header=1)

print(df.shape)
df.tail()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128380, 144)


id  member_id  \
128375                                               NaN        NaN   
128376                                               NaN        NaN   
128377                                               NaN        NaN   
128378  Total amount funded in policy code 1: 2050322100        NaN   
128379   Total amount funded in policy code 2: 819980372        NaN   

        loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
128375     5000.0       5000.0           5000.0   36 months   13.56%   
128376    10000.0      10000.0           9750.0   36 months   11.06%   
128377    10000.0      10000.0          10000.0   36 months   16.91%   
128378        NaN          NaN              NaN         NaN      NaN   
128379        NaN          NaN              NaN         NaN      NaN   

        installment grade sub_grade             emp_title emp_length  \
128375       169.83     C        C1          Payoff Clerk  10+ years   
128376       327.68     B        B3                   NaN        NaN   
128377       356.08     C        C5  Key Accounts Manager    2 years   
128378          NaN   NaN       NaN                   NaN        NaN   
128379          NaN   NaN       NaN                   NaN        NaN   

       home_ownership  annual_inc verification_status   issue_d loan_status  \
128375       MORTGAGE     35360.0        Not Verified  Oct-2018     Current   
128376           RENT     44400.0     Source Verified  Oct-2018     Current   
128377           RENT     80000.0        Not Verified  Oct-2018     Current   
128378            NaN         NaN                 NaN       NaN         NaN   
128379            NaN         NaN                 NaN       NaN         NaN   

       pymnt_plan  url  desc             purpose                    title  \
128375          n  NaN   NaN  debt_consolidation       Debt consolidation   
128376          n  NaN   NaN         credit_card  Credit card refinancing   
128377          n  NaN   NaN               other                    Other   
128378        NaN  NaN   NaN                 NaN                      NaN   
128379        NaN  NaN   NaN                 NaN                      NaN   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128375    381xx         TN  11.30          1.0         Jun-2006   
128376    980xx         WA  11.78          0.0         Oct-2008   
128377    021xx         MA  17.72          1.0         Sep-2006   
128378      NaN        NaN    NaN          NaN              NaN   
128379      NaN        NaN    NaN          NaN              NaN   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128375             0.0                    21.0                     NaN   
128376             2.0                    40.0                     NaN   
128377             0.0                    14.0                     NaN   
128378             NaN                     NaN                     NaN   
128379             NaN                     NaN                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128375       9.0      0.0     2597.0      27.3%       15.0   
128376      15.0      0.0     6269.0      13.1%       25.0   
128377      17.0      0.0     1942.0      30.8%       31.0   
128378       NaN      NaN        NaN        NaN        NaN   
128379       NaN      NaN        NaN        NaN        NaN   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128375                   f    2042.26        2042.26      3902.32   
128376                   f    3996.89        3896.97      7521.28   
128377                   w    4202.86        4202.86      8180.45   
128378                 NaN        NaN            NaN          NaN   
128379                 NaN        NaN            NaN          NaN   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
128375          3902.32          2957.74         944.58                 0.0   
128376          7333.25      

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [ ]:
# We can fix the header problem by using the 'skiprows' parameter

df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)



/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [ ]:
# Sum null values by column and sort from least to greatest
df.isnull().sum().sort_values()



inq_fi                                             2
mo_sin_old_rev_tl_op                               2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
total_cu_tl                                        2
total_rev_hi_lim                                   2
open_rv_24m                                        2
open_rv_12m                                        2
total_bal_il                                       2
open_il_24m                                        2
open_il_12m                                        2
open_act_il                                        2
open_acc_6m                                        2
tot_cur_bal                                        2
tot_coll_amt                                       2
acc_now_delinq                                     2
application_type                              

In [ ]:
# Address the extra NaNs in each column by skipping the footer as well.
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1, skipfooter=2, engine='python')



In [ ]:
print(df.shape)
df.head()

(128378, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       5000         5000           5000.0   36 months   
1 NaN        NaN      10000        10000          10000.0   36 months   
2 NaN        NaN      20000        20000          20000.0   36 months   
3 NaN        NaN       6500         6500           6500.0   36 months   
4 NaN        NaN      25000        25000          25000.0   60 months   

  int_rate  installment grade sub_grade      emp_title emp_length  \
0   11.31%       164.43     B        B3       Shipping    2 years   
1   18.94%       366.26     D        D2  Especialist 1    7 years   
2    7.56%       622.68     A        A3        Teacher  10+ years   
3   11.80%       215.28     B        B4      Educator     2 years   
4   12.98%       568.58     B        B5   Tire builder  10+ years   

  home_ownership  annual_inc verification_status   issue_d        loan_status  \
0           RENT     40000.0     Source Verified  Dec-2018            Current   
1       MORTGAGE     37115.0        Not Verified  Dec-2018            Current   
2       MORTGAGE    100000.0        Not Verified  Dec-2018         Fully Paid   
3           RENT     46500.0     Source Verified  Dec-2018  Late (16-30 days)   
4       MORTGAGE     85000.0        Not Verified  Dec-2018         Fully Paid   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN  debt_consolidation       Debt consolidation    284xx   
1          n  NaN   NaN               other                    Other    891xx   
2          n  NaN   NaN         credit_card  Credit card refinancing    982xx   
3          n  NaN   NaN  debt_consolidation       Debt consolidation    352xx   
4          n  NaN   NaN  debt_consolidation       Debt consolidation    356xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NC   9.66            0         Oct-2007               0   
1         NV  33.57            0         Jan-2002               0   
2         WA  18.92            0         Feb-1999               0   
3         AL  21.01            0         Aug-2011               1   
4         AL  36.67            1         Oct-1993               1   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN         8        0   
1                    38.0                     NaN         9        0   
2                    48.0                     NaN         9        0   
3                    61.0                     NaN        24        0   
4                    10.0                     NaN        14        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0       1070      17.5%         27                   w    2770.45   
1       5837      53.1%         15                   w    5147.99   
2      25416      29.9%         19                   w       0.00   
3       6741      41.6%         30                   w    3354.53   
4      26630      53.3%         27                   w       0.00   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0        2770.45   2956.610000          2956.61          2229.55   
1        5147.99   7309.420000          7309.42          4852.01   
2           0.00  20215.792430         20215.79         20000.00   
3        3354.53   4083.930000          4083.93          3145.47   
4           0.00  29563.763968         29563.76         25000.00   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         727.06                 0.0         0.0                      0.0   
1        2457.41                 0.0         0.0                      0.0   
2         215.79                 0.0         0.0                      0.0   
3         938.46                 0.0         0.0                      0.0   
4        4563.76                 0.0         0.0                      0.0   

  last_pymnt_d  last_pymnt_amnt next_pym

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [ ]:
df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)

print(df.shape)
df.head()

(128378, 140)


loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4      Educator     2 years           RENT     46500.0   
4     B        B5   Tire builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Sep-2020           164.43   
1         0.0                      0.0     Sep-2020           366.26   
2         0.0                      0.0     Feb-2019         20228.39   
3         0.0                      0.0     Jul-2020           215.28   
4         0.0                      0.0     Aug-2020         19356.36   

  next_pymnt_d l

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column:

In [ ]:
# Look at the first 10 values of the int_rate column
df['int_rate'].head(10)


0     11.31%
1     18.94%
2      7.56%
3     11.80%
4     12.98%
5     10.33%
6     19.92%
7     23.40%
8     15.02%
9     12.98%
Name: int_rate, dtype: object

In [ ]:
# Look at a specific value from the int_rate column
df['int_rate'][0]


' 11.31%'

In [ ]:

type(df['int_rate'][0])


str

Problems that we need to address with this column:

- String column that should be numeric
- Percent sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [ ]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them




In [ ]:
#Remove extra white space


df['int_rate'][0].strip()


'11.31%'

In [ ]:
#Doesn't work if there isn't any extra white space to remove



In [ ]:
#You can "chain" two strip functions to remove both the spaces and the % sign
df['int_rate'][0].strip().strip('%')


'11.31'

In [ ]:
# "Cast" the string value to a float
# "cast" -> Change something's data type
# This is the line of code that we're after! ->

float(df['int_rate'][0].strip().strip('%'))

11.31

### Write a function to make our solution reusable!

In [ ]:
def make_float(cell_value):
    return float(cell_value.strip().strip('%'))

In [ ]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.



In [ ]:
# Test out our function by calling it on our example
make_float(df['int_rate'][0])


11.31

In [ ]:
# is the data type correct?

type(make_float(df['int_rate'][0]))

float

### Apply our solution to every cell in a column

In [ ]:
df['float_a_boat'] = df['int_rate'].apply(make_float)

In [ ]:
# pass in *only* the name of the function, don't call it. 
# This works because we know the function works on every item in the column
# so I can simply "apply" it to the entire column

df['float_a_boat'].head()

0    11.31
1    18.94
2     7.56
3    11.80
4    12.98
Name: float_a_boat, dtype: float64

In [ ]:

# What type of data is held in our new column?

# Look at the datatypes of the last 5 columns
df.dtypes[-5:]


settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
float_a_boat             float64
dtype: object

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective 03](#pandas-apply) - Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [ ]:
# Look at the top 20 employment titles

df['emp_title'].value_counts(dropna=False).head(20)

NaN                   20943
Teacher                2089
Manager                1773
Registered Nurse        950
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Operations Manager      387
Truck Driver            387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [ ]:
# How many different unique employment titles are there currently?

len(df['emp_title'].value_counts())

43881

In [ ]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20943

What are some possible reasons as to why a person's employment title may have not been provided?

In [ ]:
# Create some examples that represent the cases that we want to clean up
import numpy as np
examples = ['owner', 'Supervisor', ' Project Manager', np.NaN]

In [ ]:
# Write a function to clean up these use cases and increase uniformity.

def clean_title(title):
  if isinstance(title, str):
    return title.strip().title()
  else:
    return "Unknown"

for example in examples:
  print(clean_title(example))


Owner
Supervisor
Project Manager
Unknown


In [ ]:
  # print things that you want to force to be shown in the notebook output
 

In [ ]:
# The value of the last thing to be returned gets printed out


In [ ]:
# if the last thing to be returned gets saved to a variable
# then it won't get printed out


In [ ]:
# if I include just a variable name, it will get printed out



In [ ]:
#Apply the function to every row of the emp_title variable



In [ ]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.
[clean_title(example) for example in examples]


['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [ ]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column
df['emp_title'] = df['emp_title'].apply(clean_title)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4       Educator    2 years           RENT     46500.0   
4     B        B5   Tire Builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Sep-2020           164.43   
1         0.0                      0.0     Sep-2020           366.26   
2         0.0                      0.0     Feb-2019         20228.39   
3         0.0                      0.0     Jul-2020           215.28   
4         0.0                      0.0     Aug-2020         19356.36   

  next_pymnt_d l

We can use the same code as we did earlier to see how much progress was made.


In [ ]:
# Look at the top 20 employment titles

df['emp_title'].value_counts(dropna=False).sort_values(ascending=False).head(20)

Unknown               20943
Teacher                2556
Manager                2395
Registered Nurse       1416
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              419
Engineer                412
Name: emp_title, dtype: int64

In [ ]:
# How many different unique employment titles are there currently?
len(df['emp_title'].value_counts())


34895

In [ ]:
# How often is the employment_title null (NaN)?

df['emp_title'].isnull().sum()

0

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [ ]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Lets look at the string format of the `issue_d` column

In [ ]:
df['issue_d'][0]

'Dec-2018'

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [ ]:
#also works for single cell ##pd.to_datetime(df['issue_d'][0])
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [ ]:
df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [ ]:
df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [ ]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128373    2018
128374    2018
128375    2018
128376    2018
128377    2018
Name: issue_d, Length: 128378, dtype: int64

Now the month.

In [ ]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128373    10
128374    10
128375    10
128376    10
128377    10
Name: issue_d, Length: 128378, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [ ]:
df['issue_month'] = df['issue_d'].dt.month
df['issue_year'] = df['issue_d'].dt.year

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4       Educator    2 years           RENT     46500.0   
4     B        B5   Tire Builder  10+ years       MORTGAGE     85000.0   

  verification_status    issue_d        loan_status pymnt_plan  \
0     Source Verified 2018-12-01            Current          n   
1        Not Verified 2018-12-01            Current          n   
2        Not Verified 2018-12-01         Fully Paid          n   
3     Source Verified 2018-12-01  Late (16-30 days)          n   
4        Not Verified 2018-12-01         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Sep-2020           164.43   
1         0.0                      0.0     Sep-2020           366.26   
2         0.0                      0.0     Feb-2019         20228.39   
3         0.0                      0.0     Jul-2020           215.28   
4         0.0                      0.0     Aug-2020         19356.36   

  next_pym

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [ ]:
df['earliest_cr_line'].head()


0    Oct-2007
1    Jan-2002
2    Feb-1999
3    Aug-2011
4    Oct-1993
Name: earliest_cr_line, dtype: object

In [ ]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], 
                                        infer_datetime_format=True)

In [ ]:
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

#Don't forget leap year = 365.25 each year.



What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [ ]:
df['days_from_earliest_credit_to_issue'].head()

0    4079
1    6178
2    7243
3    2679
4    9192
Name: days_from_earliest_credit_to_issue, dtype: int64

What's oldest credit history that was involved in Q4 2018? 

In [ ]:
df['days_from_earliest_credit_to_issue'].describe()

count    128378.000000
mean       5859.998092
std        2886.643858
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.